### PR TITLE
use vite for web builds

### PR DESCRIPTION
**Pull request by sqs**
[Vite](https://vitejs.dev/) is a fast web bundler. With Vite, a local dev server is available as fast as with esbuild, and incremental builds are (1) hot (no page reload needed) and (2) much faster (<500ms).

## Test plan

CI for e2e tests, and manually testing the dev experience:

1. Run `sg start` or whatever (I use `sg start single-program`) to get a local dev server
2. Go to https://sourcegraph.test:3443/search
3. Add some text to GlobalNavbar.tsx like `asdf1234` and hit save in your editor
4. Confirm that the edit is reflected immediately in your web browser without needing a full page reload

### CODE REVIEW

File: client/web/dev/vite/vite.config.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                },
+            },
+            modules: {
+                localsConvention: 'camelCaseOnly',
+            },
+        },
+    }
+
+    if (BAZEL) {
+        // TODO(sqs): dedupe with client/web-sveltekit
```
Comment (sqs) [inline client/web/dev/vite/vite.config.ts:76]: This TODO is not meant to be addressed yet. I think deduping it now would be premature.

cc @fkling

**Comment by sqs** [comment]
Thanks to @fkling for doing a lot of the work to get Vite to work in web-sveltekit that made this a lot easier!

**Review by eseliger — APPROVED** [review]
I don't claim to have a deep understanding of Vite, so this is mostly a stamp PR to get it rolling, but I also tried it out manually a little bit.

The rendered page update time is insanely quick with HMR, great work!

Also love that there's no 1-3s delay on page loads anymore for when esbuild starts to do its thing.

Last, having fewer proxies in the dev stack feels like a great cleanup as well, thanks for that :) 

I noticed a few things:

There seem to be no or non-working sourcemaps for scss: 

Screenshot 2023-11-24 at 20 14 34@2x

Loading a page is relatively snappy, but the amount of requests is a bit frightening 😬 
Screenshot 2023-11-24 at 20 16 02@2x
It pollutes the network tab quite badly as well. I know there's filtering, but in webpack dependencies are usually bundled together into useful chunks - is that an option for vite as well?

File: client/web/README.md [hunk 2]
```diff
@@ -29,16 +29,6 @@ To use a public API that doesn't require authentication for most of the function
 SOURCEGRAPH_API_URL=https://sourcegraph.com sg start web-standalone
 ```
 
-### Production server
```
[the remark this answers is not in the document]
Comment (sqs) [reply client/web/README.md:32]: Can't they just use the dev build? The only difference between `web-standalone-prod` and `web-standalone` is that the former is the prod build and the latter is the dev build of `client/app`. I couldn't think of a good reason to keep both, and I don't see anyone mentioning `web-standalone-prod` it in Slack, so I removed it to simplify the number of ways to run Sourcegraph in dev mode.

File: dev/Caddyfile [hunk 3]
```diff
@@ -1,20 +1,11 @@
 {
-	http_port 3081
 	auto_https disable_redirects
 }
 
-# A bit of monstrosity, since we need to reverse proxy via the web builder dev server which then
-# reverse proxies to us on HTTP.
-#
-# Caddy (tls :3443) -> web builder dev server (:3080) -> Caddy (:3081) -> sourcegraph-frontend (:3082)
+# Caddy (tls :3443) -> sourcegraph-frontend (:3080)
```
[the remark this answers is not in the document]
Comment (sqs) [reply dev/Caddyfile:5]: Yep!

**Comment by sqs** [comment]
@eseliger Fixed the dev CSS sourcemaps issue.

The number of requests being large: I looked into this, and I think this is just how Vite does it. It results in a simpler dev bundler implementation, I guess. It seems a bit slower on the initial page load, but for fast incremental reloads, I think it's a worthwhile trade-off.

**Comment by sqs** [comment]
Remaining TODOs:

- [ ] fix integration tests in Bazel on CI - they pass locally (in Bazel and without Bazel)
- [ ] fix repo symbol web e2e test in Bazel on CI - it passes locally (in Bazel and without Bazel)
- [ ] fix client/web/dev typings in Bazel on CI - it passes locally in Bazel (see below)

Logs for the last issue (see on [Buildkite](https://buildkite.com/sourcegraph/aspect-workflow-sourcegraph/builds/541#018c0839-f432-4076-99c1-29f62bfa3e5b/233-248)):
```
(20:46:03) WARNING: Remote Cache: Expected output client/web/dev/vite/manifestPlugin.d.ts was not created locally.
(20:46:03) ERROR: /mnt/ephemeral/workdir/aw-buildkite-runner-qpfv/sourcegraph/aspect-workflow-sourcegraph/client/web/dev/BUILD.bazel:18:11: output 'client/web/dev/vite/manifestPlugin.d.ts' was not created
(20:46:03) ERROR: /mnt/ephemeral/workdir/aw-buildkite-runner-qpfv/sourcegraph/aspect-workflow-sourcegraph/client/web/dev/BUILD.bazel:18:11: output 'client/web/dev/vite/vite.config.d.ts' was not created
(20:46:03) ERROR: /mnt/ephemeral/workdir/aw-buildkite-runner-qpfv/sourcegraph/aspect-workflow-sourcegraph/client/web/dev/BUILD.bazel:18:11: output 'client/web/dev/vite/manifestPlugin.d.ts.map' was not created
(20:46:03) ERROR: /mnt/ephemeral/workdir/aw-buildkite-runner-qpfv/sourcegraph/aspect-workflow-sourcegraph/client/web/dev/BUILD.bazel:18:11: output 'client/web/dev/vite/vite.config.d.ts.map' was not created
(20:46:03) ERROR: /mnt/ephemeral/workdir/aw-buildkite-runner-qpfv/sourcegraph/aspect-workflow-sourcegraph/client/web/dev/BUILD.bazel:18:11: Type-checking TypeScript project @//client/web/dev:dev_typings [tsc -p client/web/dev/tsconfig.json] failed: not all outputs were created or valid
```

**Comment by eseliger** [comment]
> @eseliger Fixed the dev CSS sourcemaps issue.

Awesome!! 

> The number of requests being large: I looked into this, and I think this is just how Vite does it. It results in a simpler dev bundler implementation, I guess. It seems a bit slower on the initial page load, but for fast incremental reloads, I think it's a worthwhile trade-off.

Yeah I kinda figured that from the docs, seems that's just how they do it. They do bundle in chunks for prod builds right? I didn't check that out locally. Seems like a fine trade-off for the massive perf gain.

**Comment by fkling** [comment]
@eseliger Basically Vite does as little work as possible during development. It avoids bundling if it can, that's why it's faster than other tools that bundle. It relies on the browser's native ability to load ES modules. It will bundle dependencies (i.e. imports from node_modules), since those usually don't change when you develop (you usually change source files).
It will bundle modules for production.

File: client/web/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhchabran) [reply client/web/README.md:32]: I'd rather re-add this if someone misses it than have a path that we're uncertain people are taking. If that becomes a problem, we'll hear about it.

File: client/web/dev/server/apiProxySettings.ts [hunk 4]
```diff
@@ -76,7 +76,7 @@ export function getAPIProxySettings(options: GetAPIProxySettingsOptions): ProxyS
         // Don't log proxy errors, these usually just contain
         // ECONNRESET errors caused by the browser cancelling
         // requests. This should not be needed to actually debug something.
-        logLevel: 'silent',
+        logLevel: 'debug',
```
Comment (jhchabran) [inline client/web/dev/server/apiProxySettings.ts:79]: I think this is a left-over?

File: client/web/dist/assets.go [hunk 5]
```diff
@@ -64,16 +66,15 @@ func (p Provider) Assets() http.FileSystem {
 		// it's already containing other files known to Bazel. So instead we put those into the dist folder.
 		// If we do detect a dist folder when running this code, we immediately substitute the root to that dist folder.
 		//
-		// Therefore, this code works with both the traditionnal build approach and when built with Bazel.
+		// Therefore, this code works with both the traditional build approach and when built with Bazel.
```
Comment (jhchabran) [inline client/web/dist/assets.go:69]: (I need grammarly in neovim 😅)

File: client/web/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/README.md:32]: Running something in prod mode on the client might be useful sometimes. Especially when we have these problems with styles ordering in production mode in different places. Maybe I missed something and there is a way to run it in production mode without sg command?

**Pull request merged**